### ISSUE TITLE
Kotlin awaiting not traced correctly.

### ISSUE THREAD

**Issue by ikovalyov** [DataDog/dd-trace-java#1123]
If you await for the deferred returned by coroutine while being in another coroutine your span information is lost and whole span is being recreated as if it was parent span. See example project:

https://github.com/ikovalyov/dd-tracing-coroutines-issue-v2

Application logs:
```
> Task :run
[dd.trace 2019-12-02 09:23:39:910 +0100] [main] INFO datadog.trace.agent.jmxfetch.JMXFetch - JMXFetch config: null [] [] [] null null {runtime-id=fedd6bdd-2499-451c-bb36-0f662f3e8648, service=unnamed-java-app} statsd:localhost:8125 System.err INFO
[dd.trace 2019-12-02 09:23:40:502 +0100] [main] INFO datadog.trace.agent.ot.DDTraceOTInfo - dd-trace - version: 0.38.0~1cb704e7
[dd.trace 2019-12-02 09:23:40:571 +0100] [main] INFO datadog.trace.agent.ot.DDTracer - New instance: DDTracer-77bb48d5{ serviceName=unnamed-java-app, writer=DDAgentWriter { api=DDApi { tracesUrl=http://localhost:8126/v0.3/traces } }, sampler=datadog.trace.agent.common.sampling.RateByServiceSampler@181d8899, defaultSpanTags={}}
[dd.trace 2019-12-02 09:23:40:600 +0100] [main] INFO datadog.trace.agent.tooling.VersionLogger - dd-trace-ot - version: 0.38.0~1cb704e7
[dd.trace 2019-12-02 09:23:40:600 +0100] [main] INFO datadog.trace.agent.tooling.VersionLogger - dd-trace-api - version: 0.38.0~1cb704e7
[dd.trace 2019-12-02 09:23:40:601 +0100] [main] INFO datadog.trace.agent.tooling.VersionLogger - dd-java-agent - version: 0.38.0~1cb704e7
2019-12-02 09:23:41.840  INFO 3998 [           main] Main$main$1                              : main DDSpan [ t_id=2654422379082193946, s_id=5753717729389507729, p_id=0] trace=unnamed-java-app/trace.annotation/Main.main metrics={} tags={component=trace, language=jvm, runtime-id=fedd6bdd-2499-451c-bb36-0f662f3e8648, thread.id=1, thread.name=main}, duration_ns=0
2019-12-02 09:23:41.892  INFO 3998 [atcher-worker-2] Main                                     : DefaultDispatcher-worker-2 DDSpan [ t_id=2654422379082193946, s_id=1179641962218689830, p_id=5753717729389507729] trace=unnamed-java-app/trace.annotation/Main.waitFor metrics={} tags={Iteration=0, component=trace, thread.id=26, thread.name=DefaultDispatcher-worker-2}, duration_ns=0
2019-12-02 09:23:41.893  INFO 3998 [atcher-worker-2] Main                                     : DefaultDispatcher-worker-2 DDSpan [ t_id=2654422379082193946, s_id=1179641962218689830, p_id=5753717729389507729] trace=unnamed-java-app/trace.annotation/Main.waitFor metrics={} tags={Iteration=0, component=trace, thread.id=26, thread.name=DefaultDispatcher-worker-2}, duration_ns=0
2019-12-02 09:23:41.899  INFO 3998 [atcher-worker-3] Main                                     : DefaultDispatcher-worker-3 DDSpan [ t_id=2654422379082193946, s_id=2170501652924201656, p_id=5753717729389507729] trace=unnamed-java-app/trace.annotation/Main.waitFor metrics={} tags={Iteration=1, component=trace, thread.id=27, thread.name=DefaultDispatcher-worker-3}, duration_ns=0
2019-12-02 09:23:41.900  INFO 3998 [atcher-worker-4] Main                                     : DefaultDispatcher-worker-4 DDSpan [ t_id=2654422379082193946, s_id=6175891178351208339, p_id=5753717729389507729] trace=unnamed-java-app/trace.annotation/Main.waitFor metrics={} tags={Iteration=3, component=trace, thread.id=28, thread.name=DefaultDispatcher-worker-4}, duration_ns=0
2019-12-02 09:23:41.899  INFO 3998 [atcher-worker-1] Main                                     : DefaultDispatcher-worker-1 DDSpan [ t_id=2654422379082193946, s_id=8485714618033939065, p_id=5753717729389507729] trace=unnamed-java-app/trace.annotation/Main.waitFor metrics={} tags={Iteration=2, component=trace, thread.id=25, thread.name=DefaultDispatcher-worker-1}, duration_ns=0
2019-12-02 09:23:41.900  INFO 3998 [atcher-worker-3] Main                                     : DefaultDispatcher-worker-3 DDSpan [ t_id=2654422379082193946, s_id=2170501652924201656, p_id=5753717729389507729] trace=unnamed-java-app/trace.annotation/Main.waitFor metrics={} tags={Iteration=1, component=trace, thread.id=27, thread.name=DefaultDispatcher-worker-3}, duration_ns=0
2019-12-02 09:23:41.902  INFO 3998 [atcher-worker-5] Main                                     : DefaultDispatcher-worker-5 DDSpan [ t_id=2654422379082193946, s_id=3463457279142380734, p_id=5753717729389507729] trace=unnamed-java-app/trace.annotation/Main.waitFor metrics={} tags={Iteration=4, component=trace, thread.id=29, thread.name=DefaultDispatcher-worker-5}, duration_ns=0
2019-12-02 09:23:41.903  INFO 3998 [atcher-worker-6] Main                                     : DefaultDispatcher-worker-6 DDSpan [ t_id=2654422379082193946, s_id=4348256562928250323, p_id=5753717729389507729] trace=unnamed-java-app/trace.annotation/Main.waitFor metrics={} tags={Iteration=6, component=trace, thread.id=30, thread.name=DefaultDispatcher-worker-6}, duration_ns=0
2019-12-02 09:23:41.904  INFO 3998 [atcher-worker-7] Main                                     : DefaultDispatcher-worker-7 DDSpan [ t_id=2654422379082193946, s_id=4421324058712740639, p_id=5753717729389507729] trace=unnamed-java-app/trace.annotation/Main.waitFor metrics={} tags={Iteration=5, component=trace, thread.id=31, thread.name=DefaultDispatcher-worker-7}, duration_ns=0
2019-12-02 09:23:41.905  INFO 3998 [atcher-worker-8] Main                                     : DefaultDispatcher-worker-8 DDSpan [ t_id=2654422379082193946, s_id=8537738308186511090, p_id=5753717729389507729] trace=unnamed-java-app/trace.annotation/Main.waitFor metrics={} tags={Iteration=7, component=trace, thread.id=32, thread.name=DefaultDispatcher-worker-8}, duration_ns=0
2019-12-02 09:23:41.906  INFO 3998 [tcher-worker-10] Main                                     : DefaultDispatcher-worker-10 DDSpan [ t_id=2654422379082193946, s_id=1985917427252727825, p_id=5753717729389507729] trace=unnamed-java-app/trace.annotation/Main.waitFor metrics={} tags={Iteration=9, component=trace, thread.id=34, thread.name=DefaultDispatcher-worker-10}, duration_ns=0
2019-12-02 09:23:41.909  INFO 3998 [tcher-worker-13] Main                                     : DefaultDispatcher-worker-13 DDSpan [ t_id=2654422379082193946, s_id=7556649444744964666, p_id=5753717729389507729] trace=unnamed-java-app/trace.annotation/Main.waitFor metrics={} tags={Iteration=8, component=trace, thread.id=37, thread.name=DefaultDispatcher-worker-13}, duration_ns=0
2019-12-02 09:23:42.942  INFO 3998 [atcher-worker-3] Main                                     : DefaultDispatcher-worker-3 DDSpan [ t_id=8401234232036504370, s_id=481145876484775946, p_id=0] trace=unnamed-java-app/trace.annotation/Main.waitFor metrics={} tags={component=trace, language=jvm, runtime-id=fedd6bdd-2499-451c-bb36-0f662f3e8648, thread.id=27, thread.name=DefaultDispatcher-worker-3}, duration_ns=0
[dd.trace 2019-12-02 09:23:43:587 +0100] [dd-trace-writer] WARN datadog.trace.agent.common.writer.DDApi - Error while sending 3 of 6 traces to the DD agent. java.net.ConnectException: Failed to connect to localhost/0:0:0:0:0:0:0:1:8126 (going silent for 5 minutes)
2019-12-02 09:23:43.946  INFO 3998 [atcher-worker-1] Main                                     : DefaultDispatcher-worker-1 DDSpan [ t_id=1567002908161829012, s_id=857742505563499203, p_id=0] trace=unnamed-java-app/trace.annotation/Main.waitFor metrics={} tags={component=trace, language=jvm, runtime-id=fedd6bdd-2499-451c-bb36-0f662f3e8648, thread.id=25, thread.name=DefaultDispatcher-worker-1}, duration_ns=0
2019-12-02 09:23:44.951  INFO 3998 [atcher-worker-7] Main                                     : DefaultDispatcher-worker-7 DDSpan [ t_id=2602817791208366767, s_id=926635854296932000, p_id=0] trace=unnamed-java-app/trace.annotation/Main.waitFor metrics={} tags={component=trace, language=jvm, runtime-id=fedd6bdd-2499-451c-bb36-0f662f3e8648, thread.id=31, thread.name=DefaultDispatcher-worker-7}, duration_ns=0
2019-12-02 09:23:45.953  INFO 3998 [atcher-worker-7] Main                                     : DefaultDispatcher-worker-7 DDSpan [ t_id=7206127938673410068, s_id=6063878620474869192, p_id=0] trace=unnamed-java-app/trace.annotation/Main.waitFor metrics={} tags={component=trace, language=jvm, runtime-id=fedd6bdd-2499-451c-bb36-0f662f3e8648, thread.id=31, thread.name=DefaultDispatcher-worker-7}, duration_ns=0
2019-12-02 09:23:46.957  INFO 3998 [atcher-worker-7] Main                                     : DefaultDispatcher-worker-7 DDSpan [ t_id=4634851895448235703, s_id=3472674439697055107, p_id=0] trace=unnamed-java-app/trace.annotation/Main.waitFor metrics={} tags={component=trace, language=jvm, runtime-id=fedd6bdd-2499-451c-bb36-0f662f3e8648, thread.id=31, thread.name=DefaultDispatcher-worker-7}, duration_ns=0
2019-12-02 09:23:47.963  INFO 3998 [atcher-worker-7] Main                                     : DefaultDispatcher-worker-7 DDSpan [ t_id=3557106430153605668, s_id=8402253980305412920, p_id=0] trace=unnamed-java-app/trace.annotation/Main.waitFor metrics={} tags={component=trace, language=jvm, runtime-id=fedd6bdd-2499-451c-bb36-0f662f3e8648, thread.id=31, thread.name=DefaultDispatcher-worker-7}, duration_ns=0
2019-12-02 09:23:48.967  INFO 3998 [atcher-worker-7] Main                                     : DefaultDispatcher-worker-7 DDSpan [ t_id=6980458062161759504, s_id=665376991271971413, p_id=0] trace=unnamed-java-app/trace.annotation/Main.waitFor metrics={} tags={component=trace, language=jvm, runtime-id=fedd6bdd-2499-451c-bb36-0f662f3e8648, thread.id=31, thread.name=DefaultDispatcher-worker-7}, duration_ns=0
2019-12-02 09:23:49.972  INFO 3998 [atcher-worker-7] Main                                     : DefaultDispatcher-worker-7 DDSpan [ t_id=4854400913329068137, s_id=7033435703755271922, p_id=0] trace=unnamed-java-app/trace.annotation/Main.waitFor metrics={} tags={component=trace, language=jvm, runtime-id=fedd6bdd-2499-451c-bb36-0f662f3e8648, thread.id=31, thread.name=DefaultDispatcher-worker-7}, duration_ns=0
```

Do you have any workaround for this?

Thanks.

**Comment by marcoferrer** [comment]
Thanks a ton for the example. I think this has to do with annotating suspend methods with `@Trace`. Suspend methods behave sequentially in kotlin but under the hood they are invoked async, almost like passing in a callback. A secondary parameter is added to method signatures and a continuation is passed in so that the coroutine can proceed with the execution of its subsequent statements. This particular use case isn’t being exercised in the existing tests at the moment but I do think its a big gap in Kotlin support. 

I’ll need to dig into the logic surrounding method level instrumentation but this example project is really going to help.

**Comment by ikovalyov** [comment]
Hi @marcoferrer 
I'm not really familiar with internal dd implementation of the tracing API, but I have few ideas on what can cause this issue.

OpenTracing API is saying that there can be only 1 active span in the thread. I guess DD java agent is using thread local to store this information.

Unfortunately in kotlin there can be multiple coroutines running in the same thread, each of them having it's own active span. Continuation logic  is processed within internal kotlin event loop where coroutine state is being checked. Also it's possible that coroutine started in one thread will be finished in another thread depending on inner logic (IO reads for example).

This is known issue and kotlin team already discussed how to achieve ThreadLocal-like functionality inside of the coroutines. The rough idea is that developer creates variable in the coroutine scope which is restored on coroutine context change.  Take a look at this discussion: https://github.com/Kotlin/kotlinx.coroutines/issues/119

There is a special CoroutineContext function designedto work with ThreadLocal variables and restoring them: https://github.com/Kotlin/kotlinx.coroutines/blob/master/kotlinx-coroutines-core/jvm/src/ThreadContextElement.kt

Unfortunately using these solutions will require to add dependency on the kotlin coroutine core to your project.

Please let me know what you think.

Thanks.

**Comment by MxKaras** [comment]
Hello @ikovalyov , thank you for the example to work on! I have opened a [pull request](https://github.com/ikovalyov/dd-tracing-coroutines-issue-v2/pull/1) to demonstrate how my company worked around this issue, using manual instrumentartion. I hope it can be of some use to you/DD.

```
> Task :run
[dd.trace 2020-02-02 21:46:19:413 +0000] [main] INFO datadog.trace.agent.jmxfetch.JMXFetch - JMXFetch config: null [] [] [] null null {runtime-id=d0ba682d-6be3-4899-bb58-1a891d737499, service=unnamed-java-app} statsd:localhost:8125
[dd.trace 2020-02-02 21:46:19:975 +0000] [main] INFO datadog.trace.agent.ot.DDTraceOTInfo - dd-trace - version: 0.42.0~f78b34fd
[dd.trace 2020-02-02 21:46:20:033 +0000] [main] INFO datadog.trace.agent.ot.DDTracer - New instance: DDTracer-51ec2856{ serviceName=unnamed-java-app, writer=DDAgentWriter { api=DDApi { tracesUrl=http://localhost:8126/v0.3/traces } }, sampler=datadog.trace.agent.common.sampling.RateByServiceSampler@714f3da4, defaultSpanTags={}}
[dd.trace 2020-02-02 21:46:20:056 +0000] [main] INFO datadog.trace.agent.tooling.VersionLogger - dd-trace-ot - version: 0.42.0~f78b34fd
[dd.trace 2020-02-02 21:46:20:056 +0000] [main] INFO datadog.trace.agent.tooling.VersionLogger - dd-trace-api - version: 0.42.0~f78b34fd
[dd.trace 2020-02-02 21:46:20:056 +0000] [main] INFO datadog.trace.agent.tooling.VersionLogger - dd-java-agent - version: 0.42.0~f78b34fd
21:46:21.322  INFO : === INIT === [ t_id=7035779805910583268, s_id=6979527234783672340, p_id=0] main}
21:46:21.464  INFO : ++++ IN +++: 5; [ t_id=7035779805910583268, s_id=3210259951591078075, p_id=6979527234783672340] DefaultDispatcher-worker-6
21:46:21.464  INFO : ++++ IN +++: 3; [ t_id=7035779805910583268, s_id=7267272080838780582, p_id=6979527234783672340] DefaultDispatcher-worker-4
21:46:21.464  INFO : ++++ IN +++: 1; [ t_id=7035779805910583268, s_id=7022831019533687639, p_id=6979527234783672340] DefaultDispatcher-worker-3
21:46:21.464  INFO : ++++ IN +++: 6; [ t_id=7035779805910583268, s_id=3530788139845839964, p_id=6979527234783672340] DefaultDispatcher-worker-7
21:46:21.464  INFO : ++++ IN +++: 9; [ t_id=7035779805910583268, s_id=7568398546412052017, p_id=6979527234783672340] DefaultDispatcher-worker-10
21:46:21.464  INFO : ++++ IN +++: 2; [ t_id=7035779805910583268, s_id=4861729612813258103, p_id=6979527234783672340] DefaultDispatcher-worker-2
21:46:21.464  INFO : ++++ IN +++: 7; [ t_id=7035779805910583268, s_id=5300620202162984080, p_id=6979527234783672340] DefaultDispatcher-worker-11
21:46:21.464  INFO : ++++ IN +++: 4; [ t_id=7035779805910583268, s_id=9080709432457601673, p_id=6979527234783672340] DefaultDispatcher-worker-5
21:46:21.464  INFO : ++++ IN +++: 8; [ t_id=7035779805910583268, s_id=814191239339242963, p_id=6979527234783672340] DefaultDispatcher-worker-9
21:46:21.465  INFO : --- OUT ---: 1; [ t_id=7035779805910583268, s_id=7022831019533687639, p_id=6979527234783672340] DefaultDispatcher-worker-3
21:46:21.464  INFO : ++++ IN +++: 0; [ t_id=7035779805910583268, s_id=5179139853462982352, p_id=6979527234783672340] DefaultDispatcher-worker-1
21:46:21.465  INFO : --- OUT ---: 0; [ t_id=7035779805910583268, s_id=5179139853462982352, p_id=6979527234783672340] DefaultDispatcher-worker-1
21:46:22.506  INFO : --- OUT ---: 2; [ t_id=7035779805910583268, s_id=4861729612813258103, p_id=6979527234783672340] DefaultDispatcher-worker-3
21:46:23.508  INFO : --- OUT ---: 3; [ t_id=7035779805910583268, s_id=7267272080838780582, p_id=6979527234783672340] DefaultDispatcher-worker-3
21:46:24.511  INFO : --- OUT ---: 4; [ t_id=7035779805910583268, s_id=9080709432457601673, p_id=6979527234783672340] DefaultDispatcher-worker-3
21:46:25.513  INFO : --- OUT ---: 5; [ t_id=7035779805910583268, s_id=3210259951591078075, p_id=6979527234783672340] DefaultDispatcher-worker-3
21:46:26.514  INFO : --- OUT ---: 6; [ t_id=7035779805910583268, s_id=3530788139845839964, p_id=6979527234783672340] DefaultDispatcher-worker-3
21:46:27.519  INFO : --- OUT ---: 7; [ t_id=7035779805910583268, s_id=5300620202162984080, p_id=6979527234783672340] DefaultDispatcher-worker-3
21:46:28.522  INFO : --- OUT ---: 8; [ t_id=7035779805910583268, s_id=814191239339242963, p_id=6979527234783672340] DefaultDispatcher-worker-3
21:46:29.526  INFO : --- OUT ---: 9; [ t_id=7035779805910583268, s_id=7568398546412052017, p_id=6979527234783672340] DefaultDispatcher-worker-3
[dd.trace 2020-02-02 21:46:31:083 +0000] [Thread-3] INFO datadog.trace.agent.common.writer.ddagent.TraceSerializingDisruptor - Flushing any remaining traces.
```

**Comment by ikovalyov** [comment]
@MxKaras thanks, I'll try this approach!

**Comment by MxKaras** [comment]
Looking back at it, our initial implementation of the `ThreadContextElement` was a bit naive. @ikovalyov if it helps -
```kotlin
class SpanContextElement(val span: Span) : ThreadContextElement<Scope> {
    override val key = Key

    // this is invoked before coroutine is resumed on current thread
    override fun updateThreadContext(context: CoroutineContext): Scope {
        return GlobalTracer.get().activateSpan(span)
    }

    // this is invoked after coroutine has suspended on current thread
    override fun restoreThreadContext(context: CoroutineContext, oldState: Scope) {
        oldState.close()
    }

    companion object Key : CoroutineContext.Key<SpanContextElement>
}
```

**Comment by ikovalyov** [comment]
Hi @MxKaras 
I approved your MR and added more testing code there. Unfortunately more issues revealed :(

Lets say we have 2 coroutines launched from parent span. They both have their own span connected to the parent span. Both of them have this SpanContextElement in use. Both of them are performing some async operations. For some reason, when first coroutine finishes while second coroutine is awaiting parent span is restored in the second coroutine instead of the correct one

```
DDSpan [ t_id=5308431498669758409, s_id=1842345218096615603, p_id=2191807972937117883] trace=Main/8/8 metrics={} tags={Iteration=8, thread.id=1, thread.name=main}, duration_ns=8064529346
15:01:48.808  INFO : --- OUT ---: 8; DDSpan [ t_id=5308431498669758409, s_id=3764630938362057172, p_id=2191807972937117883] trace=Main/launch 8/launch 8 metrics={} tags={Iteration=8, thread.id=1, thread.name=main}, duration_ns=0
DDSpan [ t_id=5308431498669758409, s_id=3764630938362057172, p_id=2191807972937117883] trace=Main/launch 8/launch 8 metrics={} tags={Iteration=8, thread.id=1, thread.name=main}, duration_ns=8061412526
--> 15:01:48.809  INFO : --- OUT ---: 9; DDSpan [ t_id=5308431498669758409, s_id=2191807972937117883, p_id=0] trace=Main/trace.annotation/trace.annotation metrics={} tags={component=trace, language=jvm, resource.name=Main.main, runtime-id=24bd7ff1-180e-4c57-8959-8e3424f75b6a, thread.id=1, thread.name=main}, duration_ns=0
--> 15:01:48.809  INFO : --- MIDDLE ---: 9; DDSpan [ t_id=5308431498669758409, s_id=2191807972937117883, p_id=0] trace=Main/trace.annotation/trace.annotation metrics={} tags={component=trace, language=jvm, resource.name=Main.main, runtime-id=24bd7ff1-180e-4c57-8959-8e3424f75b6a, thread.id=1, thread.name=main}, duration_ns=0
DDSpan [ t_id=5308431498669758409, s_id=8385449386564755215, p_id=2191807972937117883] trace=Main/9/9 metrics={} tags={Iteration=9, thread.id=1, thread.name=main}, duration_ns=9070623612
15:01:49.815  INFO : --- OUT ---: 9; DDSpan [ t_id=5308431498669758409, s_id=2191807972937117883, p_id=0] trace=Main/trace.annotation/trace.annotation metrics={} tags={component=trace, language=jvm, resource.name=Main.main, runtime-id=24bd7ff1-180e-4c57-8959-8e3424f75b6a, thread.id=1, thread.name=main}, duration_ns=0
DDSpan [ t_id=5308431498669758409, s_id=8881181702267551680, p_id=2191807972937117883] trace=Main/launch 9/launch 9 metrics={} tags={Iteration=9, thread.id=1, thread.name=main}, duration_ns=9066589061

```

**Comment by ikovalyov** [comment]
Solved it by adding 
```
    val scope = GlobalTracer.get().activateSpan(span) as TraceScope
    scope.setAsyncPropagation(true)
```
to every place where we manually create traces

**Comment by ikovalyov** [comment]
Solved it by adding the `-Ddd.trace.scope.depth.limit=0` configuration parameter

**Comment by peya-lucas-milotich** [comment]
did you try using a continuation interceptor?
https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.coroutines/-continuation-interceptor/

**Comment by tekener** [comment]
Hi @MxKaras, does the manual solution from your repo something you would recommend to use?
Thank you and that is really a shame that this is not working out of the box and needs some manual stuff to get working. 

https://github.com/ikovalyov/dd-tracing-coroutines-issue-v2/blob/master/src/main/kotlin/SpanContext.kt

**Comment by MxKaras** [comment]
My apologies @tekener - I did not notice my notifications for a long time now! I do not work at the company anymore that I was at when I wrote those messages. But the solution above was good enough for our needs (most of the time... there were a few services where the traces didn't work properly... never found out why). I think maybe that we were never creating nested spans in the way that @ikovalyov was.

Before I left that job I was looking at opentelemetry. There seemed to be some promising extensions for coroutines support, but I never got round to playing with it 🤷‍♂️ . Sorry I couldn't be more helpful!
